java/client/src/main/java/glide/api/models/GlideString.java

```diff
@@ -0,0 +1,100 @@
+/** Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.api.models;
+
+import java.util.Arrays;
+import lombok.Getter;
+
+// TODO docs for the god of docs
+public class GlideString {
+
+    @Getter private byte[] bytes;
+    private String string = null;
+
+    /** Flag whether possibility to convert to string was checked. */
+    private boolean conversionChecked = false;
+
+    private GlideString() {}
+
+    public static GlideString of(String string) {
+        var res = new GlideString();
+        res.string = string;
+        res.bytes = string.getBytes();
+        return res;
+    }
+
+    public static GlideString of(byte[] bytes) {
+        var res = new GlideString();
+        res.bytes = bytes;
+        return res;
+    }
+
+    @Override
+    public String toString() {
+        return getString();
+    }
+
+    public String getString() {
+        if (string != null) {
+            return string;
+        }
+
+        assert canConvertToString() : "Value cannot be represented as a string";
+        return string;
+    }
+
+    public synchronized boolean canConvertToString() {
+        if (string != null) {
+            return true;
+        }
+
+        if (conversionChecked) {
+            return false;
+        }
+
+        // TODO find a better way to check this
```

Comment:
You probably need to use try/catch and check for conversion failure vs trying to roundtrip the bytes.

Reply:
I have a case when conversion doesn't fail, but java patches bytes of the string.
```
bytes =
20 e6 b1 89 e5 ad 97 0b 00 24    ab    70    cc    5b    a5    28    8a    0d 0a
new String(bytes).getBytes() = 
20 e6 b1 89 e5 ad 97 0b 00 24 ef bf bd 70 ef bf bd 5b ef bf bd 28 ef bf bd 0d 0a
```
in this examples some bytes were replaced - it is a data corruption.